2024 blog post on vgaterm (#23)
dougli1sqrd/eric-off-the-rails: static/_posts/2024-04-01-vgaterm-intro-001.md

---
layout: authored-post
title: "Introduction to Vgaterm"
author: Eric
category: tech
tags: vgaterm
feature_image: "/assets/images/vgaterm-banner-01.jpeg"
---

> Over the past two years or so my friend [Seth](https://github.com/sethp) and I have been building a hardware [terminal](). Or at least, we started building something and it eventually became a terminal.

Reviewer: Broken link?

> This pandemic project was a way for us to connect with each other and learn new technical skills while building something that would allow us to physically connect computers together.

Reviewer: <3

> A hardware terminal is an old-school way of connecting a user to a mainframe - or a multi-user computer. A terminal would have a keyboard and monitor and a "serial" connection to a computer that many other people could be using at the same time on their own terminals. It's the hardware equivalent to opening your favorite terminal program and typing the `ls` command to list all the files in the current directory. To learn about terminals and how they've shaped computing history, I'd absolutely reccomend [this blog post](http://www.linusakesson.net/programming/tty/)

Reviewer: I think the comparison to a terminal emulator program (software)  is good, but I'm not sure it's super clear the link between `ls` and the physical hardware device here. 

Maybe something like:

> It's the physical device that you might've had on your desk in 1975, and that now you have as many software-emulated copies of today as you'd like via your favorite terminal program.

worked into the paragraph?

> Terminals are largely not in use any more because computers are cheap and everywhere and can just connect to each other over the internet. You may remember using a hardware terminal (up to the late 90s, perhaps) at your local library to browse the available books in the catalogue.

Reviewer: Oh, is this something you experienced? That's cool: maybe you can say so?

This is the start in a series of posts all about the planning, building, and philosphy of Seth and I making a hardware terminal in 2024.

This post will specifically be about the basic workings, pictures, and how we were inspired to make something like this.

# The Start

> My technical background is mostly java, python, and the web. So a part of me has always wanted to learn more systems and lower-level programming. When Rust came around I was excited to learn a new systems language with an interesting type system. At some point many years back during a conversation over pizza, Seth and I decided to make a Github org for playing with Rust and learning low level programming. We called it [Rustbox](), as the original conceit was to make an operating system together in rust.

Reviewer: Link here too?

> At some point I learned about [RISC-V](https://riscv.org/), an open source instruction set for CPUs, and decided I wanted to try programming rust for RISC-V, and Seth and I bought a few [HiFive](https://www.sifive.com/boards/hifive1-rev-b) boards. Inspired by [Ben Eater](https://www.youtube.com/@BenEater)'s youtube video ["Making the Worlds Worst Video Card"]() using a 6502 processor, I decided to make the hifive board into a "video card". It output a VGA signal in the 640x480 mode with 160x96 actual pixels (due to memory constraints) with 64 colors. After a year I had a set of breadboards and a RISC-V CPU displaying images and text (using my own font):

Reviewer: Link?

Reviewer: Also, these last two sentences flow kinda awkwardly together. Maybe something like:

> ... I decided to make the hifive board into a "video card". After a year I had a set of breadboards and a RISC-V CPU (see below) displaying images and text using my own font. It output a VGA signal in the 640x480 mode made up of 160x96 actual pixels, each with 64 possible colors (due to memory constraints).

Reviewer: Oh, and maybe consider adding some preamble saying "hey I don't intend to talk much about this version" so it's clear it's not going to have the full level of detail as vgaterm?

{% include figure.html image="/assets/images/hifive-vga-breadboards.jpg" caption="hifive board attached to breadboards outputing VGA" width=600 %}

{% include figure.html image="/assets/images/streetcar-red-hifive.jpg" caption="Red streetcar in SF converted for VGA hardware" width=600  %}

{% include figure.html image="/assets/images/green-text-hifive.jpg" caption="Text demo with highlighted characters included using my own custom font" width=600 %}

Code for this project can be [found here](https://github.com/rustbox/rustbox-hifive1-revb)

# Vgaterm

> The hifive board video card was neat and I learned a lot but I wanted it to be useful somehow. In the hifive code I had also made some command line bare metal programs as well as having the characters update from keyboard input. I was also always a little sad that I needed a large computer to be attached to my small computer to run and interact with it (aka through screen or other terminal program). Thus I realized I wanted to make a physical terminal, and upgrade the VGA capabilities. So Seth and I embarked on making *Vgaterm*: a hardware terminal that would would output 640x400 pixels at 256 colors, connect to serial IO, recieve keyboard input, and display text updates to the screen. The goal was to be able to log in to a standard linux machine with the hardware and interact on the command line - and hopefully use any standard terminal programs like Vim, etc.

Reviewer: I'm not sure I follow this sentence:

> In the hifive code I had also made some command line bare metal programs as well as having the characters update from keyboard input.

Maybe strike it out? The "I wanted to have a little computer" feels like enough justification.

Reviewer: Also, this feels like maybe a good time to introduce our hubris: "we do this not because it was easy, but because we _thought_ it would be easy" lol

Like, "We picked what we thought was a _modest_ target: *full* 640x480 resolution, a mere 256 colors ... "?

Reviewer: Still not entirely sure what that second sentence is saying, here.

{% include figure.html image="/assets/images/vgaterm-top.jpg" caption="Completed Vgaterm" width=750 %}

{% include figure.html image="/assets/images/vgaterm-front.jpg" caption="Vgaterm uses a four position rotary switch to select an IO port: USB, raw pin headers (pictured here bridged), RJ45, DSUB-9 serial (RS-232)" width=750 %}

{% include figure.html image="/assets/images/vgaterm-hello.jpg" caption="Hello World, this is the Vgaterm Terminal!" width=750 %}

## Technical Overview

Since the new device aspirationally would output VGA at the full 640x480 60Hz standard we would need a microcontroller that would:

* have at least 300 KB of fast memory for the video buffer (1 byte per pixel, giving us 256 colors)
* have some memory for coordinating IO and other tasks
* fast enough to output pixels for the VGA pixel clock at 25.175 MHz: 39.7 ns per byte

Additionally, part of the philosophy of the project was to celebrate and engage in open hardware and software. So keeping a RISC-V core felt crucial to our goals.

Finding the esp32c3 by Espressif seemed to fit the bill by the numbers:
* 400 KB SRAM
* 4 MB Flash Memory
* 160 MHz CPU
* 15 GPIO pins

> However, it turned out that bit-banging 8 GPIO lines as fast as possible on the esp32c3 was not quite fast enough, additionally there wasn't an obvious way to match the VGA clock domain from the CPU.

Reviewer: ```suggestion
However, it turned out that bit-banging 8 GPIO lines as fast as possible on the esp32c3 was not quite fast enough. Additionally, there wasn't an obvious way to match the VGA clock domain from the CPU: <provide a one-sentence description as to why>
```

> To solve the clock domain issue we decided to use a FIFO memory (first in first out, like a queue) chip. So the CPU writes bytes into the FIFO, and the VGA clock domain reads from the FIFO, 

Reviewer: To solve the clock domain issue, "and buy ourselves some margin" / "engineering tolerances" maybe?

There's also a trailing comma here, but it seems like a complete thought on its own.

Reviewer: Maybe mention that FIFOs are hardware designed for exactly this kind of clock domain crossing (optionally: because of "metastability" issues)?

To solve the speed issue the esp32c3 comes equipped with a quad [SPI](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface) module allowing 4 bits to be clocked out at up to 80 MHz - much faster than bit-banging.

So with these two pieces, we could design a circuit that collects two 4 bit [nibbles](https://en.wikipedia.org/wiki/Nibble) into a byte (representing the color of a single pixel) and load it into a FIFO memory. At the same time, the VGA timing hardware would read pixel bytes exactly when needed.

> 

Reviewer: I think it's definitely worth mentioning the constraint that the FIFO was too expensive to hold the whole frame buffer: "We asked each other why not put the entire frame into the FIFO? Or store the whole frame buffer off chip? And the short answer is that it was too expensive: more details in a future post ..."

### Logical Components

There are five main components to Vgaterm:

* VGA Timing
* CPU Pixel Output
* Displaying a Pixel Byte
* Digital to Analog Converters (DAC)

> * IO control and video buffer updates

Reviewer: AKA "the terminal parts" lol

{% include figure.html image="/assets/images/vgaterm-logical-diagram.jpg" caption="Vgaterm logical diagram. Details were omitted to show the high level units" width=800 %}

Repositories containing all our code and schematics:
* esp32c3 code: [https://github.com/rustbox/esp32c3-vgaterm](https://github.com/rustbox/esp32c3-vgaterm)
* VGA timing generation code and Programmable Logic Device code: [https://github.com/rustbox/cuddly-robot](https://github.com/rustbox/cuddly-robot)
* PCB schematics and CAD: [https://github.com/rustbox/vgaterm-schematics](https://github.com/rustbox/vgaterm-schematics)

### VGA Timing

> 

Reviewer: Maybe add an intro here about what role the VGA timing component plays; you can probably get a good motivation for it by answering "what happens if the VGA timing part stops working?"

> VGA comes in several modes, and we target the 640x480, 60Hz frame rate mode - the base standard. VGA has 3 analog signals between 0 and 0.7 volts representing the amount of Red Green and Blue per pixel. There are also two timing signals `H_SYNC` and `V_SYNC` that tell the monitor what resolution mode to be in and when to sample the analog color signals to actually dispaly a pixel. When the monitor is actively sampling colors and displaying pixels we say that this portion of the frame is `visible`. There is time during a complete frame when this isn't the case and this is _blanking area_. Lastly, the _pixel clock_ speed is the rate at which pixels are shown, and in our case that's 25.175 MHz. In the diagram above, I've labeled a signal `CLK_VIS`, and this is the _pixel clock_ signal during the `visible` time. To see detailed VGA timing information, see [tiny vga](http://tinyvga.com/vga-timing/640x480@60Hz).

Reviewer: I'd suggest revisiting this section. Like, it's all good definitions of terms, but I might save most of it for the VGA timing in-depth blog post: for this post I think it's enough to just say "we need some sync signals that define the frame", like the next paragraph does.

Author: My feeling is that we def need some discussion of what the VGA signal is in the overview. This still very much glosses over the details it seems like to me.

Author: I guess especially since I have that diagram, I wanted to make sure that the diagram is explained a little

Reviewer: I agree that we ought to set up the sort of "seams" we used to break the system apart, I don't think this paragraph does that well. The next one does a much better job, IMO.

What feels especially important to include about the VGA spec, do you think?

:----:|:-----:

> ![](/assets/images/vgaterm-timing-front.jpg) | ![](/assets/images/vgaterm-timing-back.jpg)

Reviewer: alt texts?

<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>

> VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.

Reviewer: Oh, probably not for this blog post, but I'm curious if you happen to have the size of the timing window we bought ourselves with the FIFO in your notes anywhere? And where we placed `START` relative to it?

### Pixels Out

> Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. This is the most important software role as any jitter in timing can cause display artifacts.

Reviewer: ```suggestion
Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. Too much jitter in the software timing will cause display artifacts.
```

Reviewer: Bury the lede on this one, mebe?

Author: oh are you saying that I did bury the lede (why isn't this spelled lead, it's never made sense to me lol) and I shouldn't or that I didn't bury the lede and I should?

Reviewer: apparently it's spelled lede (guide) specifically to distinguish it from lead (Pb), which was also a term of art for spacing the type: https://www.merriam-webster.com/wordplay/bury-the-lede-versus-lead

Reviewer: When I read the "most important software role" I said "wait, is it? what about ....", so, here, I'm saying I think you ought to omit the "most important job" detail. This presents it as a ground-truth fact akin to "the VGA spec mandates a 21.175MHz clock," but it's not like that at all.

You do a better job talking to it later on as a decision we made, and so I think the decision here is to either present the problem (visual artifacts) and bury the lede (that we decided avoiding those were Most Important), or—as an alternative—foreshadow the decision making process by saying something more like:
> ... to hardware so they may be displayed. We decided this is the most important software role as any jitter in timing can cause display artifacts << and also see below for more details on that decision >>

> To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
> 
> This was clever, but there are a few constraints we had to consider when designing a circuit around the FIFO:

Reviewer: ```suggestion
Our "chasing-the-beam" FIFO approach was clever, but there are a few constraints we had to consider when designing a circuit around the FIFO:
```

(and then also make sure to describe the "chasing the beam" bit here or earlier)

Author: ha is this just so I say chasing the beam?

Reviewer: In that RTL class I took, I learned (via, like, offhand kind of references) that putting FIFOs between clock domains is the standard approach. So "to cross the clock domains, we cleverly used a FIFO" reads a little like "to determine the area of the rectangle, we cleverly multiplied the length times the width": it's extremely clever, which is why 1) we stumbled into it from first principles, and 2) everyone uses it so much that it's kind of banal. 

I think there ought to be just a little more detail as to what was clever/constraining about our approach: what we were reacting to (the price of memory & difficulty in constructing a full off-chip VRAM bank, especially on breadboards?), and/or a little more flavor as to how difficult the challenge of precise hardware/software synchronization first appeared (can we even make the software be cycle-precise? wait, are the clock cycles even precisely the same between chips? etc...) and thus how clever we felt when we said "wait, let's buffer it!".

Both is good too—but you gotta save something for the future posts, yeah?

> * The FIFO read and write setup and hold times had to be obeyed by the SPI clock and `CLK_VIS` signals

Reviewer: ```suggestion
*The data going into the FIFO had to obey the setup and hold times within each SPI clock cyle
```

Author: can you say a little more for the motivation for the suggestion here? Both the read and the write side have to obey their respective setup and hold times

Reviewer: Yeah, there was a stylistic and a technical change compacted into that one suggestion. Splitting them apart:

1. (stylistic) The sentence had been written with a voice that was passive, and when I did that in high school I got angry red pen marks from a fierce little italian man and now I reflexively rewrite passive voice sentences to active. (Actually, did you have the same English teacher as Heather? She does the same thing all the time too)

2. (technical) It feels weird to say that the read/write setup & hold times are a property of the clocks, rather than the data relative to the clock. There is a constraint there, sure, but we added the setup + hold times together and selected a FIFO that was capable of being driven by a 20MHz write / 21.175MHz read signal. I think we even did that on the first try? The much more interesting part (that we struggled with) was how to mesh the data, clock, and up/downstream circuits. 

Maybe you could say a little more about what you want to set up here, and what you hope to leave for the "timing analysis, a journey and erroneous comedy" subsection of whatever in-depth post?

Reviewer: Looks like this comment chain got marked as outdated, but the bullet point's wording didn't change.

> * The SPI Clock could be in only divisions by 2 of 80 MHz (80, 40, 20, 10, etc)

Reviewer: ```suggestion
* The SPI only supported clock speeds in power-of-two divisions of 80 MHz (80, 40, 20, 10, etc)
```

* If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels

>     * This could be mitigated by adding more space in the FIFO, but at a steep dollar cost

Reviewer: ```suggestion
```

(and cover the dollars-to-safety-margin tradeoff earlier)

Author: I was thinking the first section where you had the comment about cost was to just introduce everything and this section would go into a little more detail where the cost would be talked about along with the other constraints

Reviewer: It's just that this was the only sub-bullet in the list, and I don't know that it justified its existence. Especially given it's probably worth adding this much context when we're introducing the FIFO, which happened earlier in the "Technical Overview".

If you did want to go more into detail here, that would be great: I'd love to see e.g. "the FIFO was $20, which was 5-20x the price of all our other ICs, so we only had a couple from when we just wild-ass guessed that 4k would be a good size, and we didn't want to buy bigger ones" or some sort of a retroactive justification analysis in terms of dollars/µs for a 4k vs 8k FIFO

* If we emitted pixels too slowly, and VGA hardware attempted to consume pixels faster than were being written then we'd also drop pixels
* VGA hardware does not read from the FIFO at the pixel clock rate across the entire duration of the frame. Since the visible portion of the frame is only about 73% of the total time spent per frame (because of the blanking times), the average read rate per frame is actually less than the pixel clock rate

> The details in how we eneded up on our specific design will be elaberated on in another post, but suffice to say we had to coordinate design details across software, hardware, specific part choices, and part pricing.

Reviewer: :man_cook: :pinched_fingers:

### Displaying Pixels
To dispaly pixels, the `CLK_VIS` signal generated by the VGA timing hardware reads bytes from the FIFO and into a Programmable Logic Unit (PLU) which maps the 8 bits into 9 bits, 3 sets of 3 bits for each of the color channels - red, green, and blue. These 9 bits then get sent to the DAC, described in the following section.

An important consideration for this stage is how long it takes to shuttle a byte from the FIFO into the DAC and finally into the monitor. Since there is processing of the read bytes it's possible that the pixels become late compared to `H_SYNC` and `V_SYNC` and this will show up as a horizontal shift. I.E. the "first" pixel could actually show up on the 2nd or 3rd pixel position on the screen, instead of the actual 1st pixel.

{% include figure.html image="/assets/images/left-collum-timing-bug.jpg" caption="Notice the far left column of the screen in white next to the blue. It appears to wrap around from the far right side of the screen because the first pixel of the image is shifted over 2 or 3 positions" width=550 %}

> So a method to handle this kind of pixel shifting makes the picture quality better.

Reviewer: ```suggestion
So we developed a method to handle this kind of pixel shifting to synchronize the pixel hardware with the VGA timing hardware.
```

> ### Digital to Analog Converter (DAC)

Reviewer: I really like this section: it feels like you really hit your stride on motivating the "why" & then providing the "what" at a real tantalizing teaser kind of way.

> Once the pixel byte has been taken out of the FIFO memory it's on its way to being displayed by the monitor. To be displayed this pixel byte must be converted into three analog voltages between 0 and 0.7 volts corresponding to the Red, Green, and Blue color channels. This is accomplished with a [R2R Resistor Ladder DAC]().

Reviewer: "This is accomplished" -> "We accomplished this"—there are many and more ways to do DACs, and ours is cheap & easy & simple but certainly not exclusive of the other options.

Reviewer: Also, linky?

> An R2R Resistor Ladder DAC is traditionally connected into an op-amp to maintain the output voltage regardless of load. However in our case the load is just the 75 Ohm impedance of the analog color lines in the monitor, so we really only have the one load. One concern, however, is that the resistor ladder can consume a fair amount of current and the lines representing the color coming from an IC may not be rated to supply the amount of current the DAC would need to maintain the voltage.

Reviewer: ```suggestion
A stand-alone DAC is commonly terminated in an op-amp stage to compensate for output impedance and maintain the output voltage regardless of what load it's connected to. However in our case the load is just the 75Ω impedance of the analog color lines in the monitor, so we opted to save on the cost of an op-amp and instead account for the impedance in our DAC design. One concern, however, is that the resistor ladder will consume a fair amount of current, and the lines representing the color coming from an IC may not be rated to supply the amount of current the DAC would need to maintain the voltage.
```

Reviewer: Also maybe mention why we didn't use an off-the-shelf DAC? ("we wanted to learn how to build a DAC" is a fine reason, even if it's kind of revisionist)

> In earlier versions of the DAC this is what we did (this is what Ben Eater did in his original "Worlds Worst Video Card" video). We did also try an op-amp but that seemed to create some weird artifacts in the colors in images. Lastly what seemed to work best is just feed the input bit lines into line drivers which _could_ supply the appropriate amount of current and now colors are quite vivid.

Reviewer: Maybe save this part for the DAC post, too?

### IO Control

> To actually function as a terminal Vgaterm must receive serial signals from a connected host on one of the four IO ports and update the video buffer accordingly. It also must handle keyboard input and forward those typed characters to the connected host.

Reviewer: Four?

Author: USB, pin headers, RJ45 (RS-232), and DSUB-9 Serial (RS-232). Are you saying I should put that here too?

Reviewer: yeah, I'd say either qualify or omit the number: also maybe call 'em the "terminal IO" ports or something? Since VGA and keyboard are also both IO ports

{% include figure.html image="/assets/images/vgaterm-terminal-sequence.drawio.png" caption="Basic Keyboard input to terminal operation" %}

> For keyboard input we went over several options including falling back on [PS/2 protocol over USB](https://en.wikipedia.org/wiki/PS/2_port#Conversion_between_PS/2_and_USB) and trying to write a basic USB host onto the esp32c3. But we settled on just using this [USB host to UART converter](https://www.tindie.com/products/matzelectronics/ch559-usb-host-to-uart-bridge-module/) by [MatzElectronics](https://www.tindie.com/stores/matzelectronics/).

Reviewer: ```suggestion
For keyboard input we went over several options including falling back on [PS/2 protocol over USB](https://en.wikipedia.org/wiki/PS/2_port#Conversion_between_PS/2_and_USB) and trying to write a basic software-driven USB host with the esp32c3 GPIO pins. But we settled on using this [USB host to UART converter](https://www.tindie.com/products/matzelectronics/ch559-usb-host-to-uart-bridge-module/) by [MatzElectronics](https://www.tindie.com/stores/matzelectronics/).
```

Software, in a loop, waits for keyboard input or input by a connected host. When input is received on either, the CPU is interrupted. If receiving key presses, the character(s) are dequeued and passed along to the connected host. If receiving characters from the host, vgaterm updates the text buffer and cursor which updates the video memory. The updated video memory is then emitted in the next frame, showing the text to the user on screen.

> Seth and I decided that this should _never_ interrupt the CPU from emitting pixel bytes as we wanted to prevent any visual artifacts as much as possible. However this means that we could end up more likely to drop characters if the CPU is overwhelmed with input. The CPU UART is configured to issue an `XOFF` character as a part of [Software Flow Control](https://en.wikipedia.org/wiki/Software_flow_control) to tell the host to stop sending input.

Reviewer: I'd suggest developing this a little bit more on the problem/tradeoff side, and maybe defer XOFF for the follow-up post.

Some important points that feel worth hitting:

* there's hardware buffering & other things that reduce the likelihood of drops on the UART stuff that doesn't exist on the video side (because we didn't build it) 
* frames are ephemeral, and bytes are "forever," but screen shake feels way more like jank (especially given the relative frequencies)
* we uncovered a big difference between average and worst-case terminal processing: and we on purpose decided against over-building for the worst case

> # Next Steps

Reviewer: ```suggestion
# Closing Thoughts
```

Another aspect about this project that feels important is how we engaged with open source and open hardware. Seth and I wanted this to be open from the beginning, but I feel like I have a love for open source and open hardware like never before.

Engaging with the open hardware world helped me realize that electronics and hardware is just a skill that someone can learn, and that we, the community, can collectively own and build things that are important and useful - that we don't have to capitulate to huge corporations that want to collect our data, exploit workers, or monopolize the objects and services in our daily lives. It feels like a way to fight against the [enshittification](https://en.wikipedia.org/wiki/Enshittification) that represents our time in tech today. These are concepts I'm still thinking about and working through, and I have Vgaterm to thank for helping me engage with these ideas.

> 

Reviewer: <3 big :+1: from me on this section :smile:

---
layout: authored-post
title: "Introduction to Vgaterm"
author: Eric
category: tech
tags: vgaterm
feature_image: "/assets/images/vgaterm-banner-01.jpeg"
---
{% include figure.html image="https://upload.wikimedia.org/wikipedia/commons/9/9f/DEC_VT100_terminal_transparent.png" width=400 position="right" caption="DEC VT100 terminal" %}

Over the past two years or so my friend [Seth](https://github.com/sethp) and I have been building a hardware [terminal](https://en.wikipedia.org/wiki/Computer_terminal). Or at least, we started building something and it eventually became a terminal!

This pandemic project was a way for us to connect with each other and learn new technical skills while building something that would allow us to physically connect computers together.

> A hardware terminal is an old-school way of connecting a user to a mainframe - or a multi-user computer. A terminal has a keyboard and monitor and a "serial" connection to a computer that many other people could be using at the same time with their own terminals. It's the physical device that you might have had on your desk in 1975, and that now exist as software emulating that hardware in the form of your favorite terminal program where you might type `ls` to list files in a directory. To learn about terminals and how they've shaped computing history, I'd absolutely reccomend [this blog post](http://www.linusakesson.net/programming/tty/).

Reviewer: ```suggestion
A hardware terminal is an old-school way of connecting a user to a mainframe - or a multi-user computer. A terminal has a keyboard and monitor and a "serial" connection to a computer that many other people could be using at the same time with their own terminals. It's the physical device that you might have had on your desk in 1975, and that now exists as software emulating that hardware. That's why your favorite program where you might type `ls` to list files in a directory is called a "terminal": it's short for "terminal emulator"! To learn about terminals and how they've shaped computing history, I'd absolutely recommend [this blog post](http://www.linusakesson.net/programming/tty/).
```

exist -> exists
reccomend -> recommend

(not sure if it's worth throwing this into a spellchecker somewheres?)

Then, consider breaking up the sentence: I like the "you're still using it today!" but it's a lot to go from 1975 to "typing ls to list files in directory."

> Terminals are largely not in use any more because computers are cheap and everywhere and can just connect to each other over the internet. You may remember using a hardware terminal (up to the late 90s, perhaps) at your local or school library to browse the available books in the catalogue. I remembery my elementary school likely used VT terminals in their library!

Reviewer: ```suggestion
Terminals are largely not in use any more because computers are cheap and everywhere and can just connect to each other over the internet. They could still be found up to the late 90s, though: I remember using a hardware terminal at my elementary school library to browse the available books in the catalogue! I think it was even a VT terminal, since they were such a popular device.
```

That sentence feels a little like at least two sentences in a bag yowling to be let out. It's a fun detail to share, but I think it comes across more clearly on its own.

> This is the start in a series of posts all about the planning, building, and philosophy of Seth and I making a hardware terminal in 2024.

Reviewer: ```suggestion
This is the start in a series of posts all about the planning, building, and philosophy of making a hardware terminal in 2024.
```

> This is the start in a series of posts all about the planning, building, and philosophy of Seth and I making a hardware terminal in 2024.
> 
> This post will specifically be about the basic workings, pictures, and how we were inspired to make something like this.

Reviewer: ```suggestion
This is the start in a series of posts all about the planning, building, and philosophy of making a hardware terminal in 2024. This post will specifically be about the basic workings and inspiration to make something like this. I'll share a few of the challenges we ran into, and plenty of pictures!
```

I think this flows a little better.

# The Start

> My technical background is mostly java, python, and the web. So a part of me has always wanted to learn more systems and lower-level programming. When Rust came around I was excited to learn a new systems language with an interesting type system. At some point many years back during a conversation over pizza, Seth and I decided to make a Github org for playing with Rust and learning low level programming. We called it [Rustbox](https://github.com/rustbox), as the original conceit was to make an operating system together in rust.

Reviewer: ```suggestion
My technical background is mostly java, python, and the web. So a part of me has always wanted to learn more systems and lower-level programming. When Rust came around I was excited to learn a new systems language with an interesting type system. At some point many years back during a conversation over pizza, Seth and I decided to make a Github org for playing with Rust and learning low level programming. We called it [Rustbox](https://github.com/rustbox), as the original concept was to make an operating system together in rust.
```

I like conceit, but it does sort of mean that we were overly confident:

![conceit definition](https://github.com/dougli1sqrd/eric-off-the-rails/assets/241129/bec54b08-5f0e-4503-a4a6-439826a6d8ff)

Which, I mean, we were!

> At some point I learned about [RISC-V](https://riscv.org/), an open source instruction set for CPUs, and decided I wanted to try programming rust for RISC-V, and Seth and I bought a few [HiFive](https://www.sifive.com/boards/hifive1-rev-b) boards. Inspired by [Ben Eater](https://www.youtube.com/@BenEater)'s youtube video ["Making the Worlds Worst Video Card"](https://www.youtube.com/watch?v=l7rce6IQDWs) using a 6502 processor, I decided to make the hifive board into a "video card". After a year I had a set of breadboards and a RISC-V CPU displaying images and text using my own font. It output a VGA signal in the 640x480 mode made up of 160x96 actual pixels, each with 64 possible colors (due to memory constraints):

Reviewer: ```suggestion
At some point I learned about [RISC-V](https://riscv.org/), an open source instruction set for CPUs, and decided I wanted to try programming rust for RISC-V, and Seth and I bought a few [HiFive](https://www.sifive.com/boards/hifive1-rev-b) boards. Inspired by [Ben Eater](https://www.youtube.com/@BenEater)'s youtube video ["Making the Worlds Worst Video Card"](https://www.youtube.com/watch?v=l7rce6IQDWs) using a 6502 processor, I decided to make the hifive board into a "video card". After a year I had a set of breadboards and a RISC-V CPU displaying images and text using my own font. It output a VGA signal in the 640x480 mode, that was made up of 160x96 actual pixels, each with 64 possible colors due to memory constraints:
```

The final device uses an esp32c3 microcontroller and custom designed circuit boards hand soldered together.

{% include figure.html image="/assets/images/vgaterm-top.jpg" caption="Completed Vgaterm" width=750 %}

{% include figure.html image="/assets/images/vgaterm-front.jpg" caption="Vgaterm uses a four position rotary switch to select an IO port: USB, raw pin headers (pictured here bridged), RJ45, DSUB-9 serial (RS-232)" width=750 %}

{% include figure.html image="/assets/images/vgaterm-hello.jpg" caption="Hello World, this is the Vgaterm Terminal!" width=750 %}

## Technical Overview

Since the new device aspirationally would output VGA at the full 640x480 60Hz standard we needed a microcontroller that would:

* have at least 300 KB of fast memory for the video buffer (1 byte per pixel, giving us 256 colors)
* have some memory for coordinating IO and other tasks

> * fast enough to output pixels for the VGA pixel clock at 25.175 MHz: 39.7 ns per byte

Reviewer: ```suggestion
* be fast enough to output pixels for the VGA pixel clock at 25.175 MHz: 39.7 ns per byte
```

Additionally, part of the philosophy of the project was to celebrate and engage in open hardware and software. So keeping a RISC-V core felt crucial to our goals.

Finding the esp32c3 by Espressif seemed to fit the bill by the numbers:
* 400 KB SRAM
* 4 MB Flash Memory
* 160 MHz CPU
* 15 GPIO pins

> However, it turned out that bit-banging 8 GPIO lines as fast as possible on the esp32c3 was not quite fast enough because of limitations on the device. Additionally there wasn't an obvious way to match the VGA clock domain from the CPU. To get the pixel timing right we would need to time out intervals of 39 ns - the period of the VGA pixel clock. At 160 MHz CPU clock speed that's only 6 cycles per pixel period. Additionally we'd need the timing for the other VGA signals (`VSYNC` and `HSYNC`) to be quite precise and it just didn't seem feasible to generate those timings within the CPU.

Reviewer: Do you want to say which limitations here? IIRC it was the whole clock domain shift that meant the GPIO stuff couldn't be driven at anywhere near the 160MHz clock (each CSR write that would toggle the pin would take dozens or so of cycles to complete)

Reviewer: ```suggestion
However, it turned out that bit-banging 8 GPIO lines as fast as possible on the esp32c3 was not quite fast enough because of limitations on the device. Additionally there wasn't an obvious way to match the VGA clock domain from the CPU. To get the pixel timing right we would need to time out intervals of 39 ns - the period of the VGA pixel clock. At 160 MHz CPU clock speed that's only 6 cycles per pixel period. Finally, we'd need the other VGA timing signals (`VSYNC` and `HSYNC`) to be quite precise and it just didn't seem feasible to generate those timings within the CPU.
```

E_TOO_MANY_ADDITIONALLYs

Additionally, this paragraph's flow into the next felt a little choppy to me: I *think* moving the "timing" around helps clarify that there's "clock domain" vs, like, signal-framing or VGA-bounding "timing" as two separate concepts, so it doesn't read like we're presenting the FIFO as a solution to the latter.

> To solve the clock domain issue we decided to use a FIFO memory (first in first out, like a queue) chip. So the CPU writes bytes into the FIFO, and the VGA hardware reads from the FIFO at a different rate. As long as the CPU can write bytes within a certain tolerance, the CPU and the VGA display hardware needs to only be loosely coupled.

Reviewer: It'd be nice to say what the (looser) coupling is rather than just asserting there's "less coupling" (which is sort of software-speak for "doubleplusgood" :upside_down_face:). Also, this feels like the right time to introduce how much the FIFO tail wagged the whole dog, and a little bit of why: the FIFO thing ended up being our most interesting choice! and hardest to un-do! That's a key take-away there, I think.

Maybe something like:

```suggestion
To solve the clock domain issue we decided to use a FIFO memory (first in first out, like a queue) chip. So the CPU writes bytes into the FIFO, and the VGA hardware reads from the FIFO at a different rate. The size of the FIFO buys us synchronization buffer: a larger FIFO would increase the tolerance on when the CPU can write bytes relative to when the VGA hardware needs to read them. Selecting a FIFO ended up being a primary design constraint, though: there were only a few sizes available, and even at a relatively small 4k of memory the FIFO was still our most expensive single component at $20 (10-100x more expensive than the other ICs).
```

> To solve the speed issue we utilized the esp32c3 quad [SPI](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface) module allowing 4 bits to be clocked out at up to 80 MHz - much faster than bit-banging. Our actual implementation ended up using a 40 MHz SPI clock speed.

Reviewer: The 80MHz thing feels a little bait-and-switchy? Like, why _didn't_ we use the 80MHz?

```suggestion
To solve the speed issue we utilized the esp32c3 quad [SPI](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface) module allowing us to clock out 4 bits to be clocked out at 40 MHz - much faster than bit-banging. In order to emit a whole frame over SPI at that speed, we also had to learn how to use DMA on the software side.
```

Reviewer: You might add something like "the hardware was capable of faster clock rates, but we couldn't take advantage of it"; IIRC the two main reasons were:

* software: interrupt latency!
* hardware: breadboards!!!

> Lastly, even with the faster SPI pixel output, emitting the full 640x480 frame proved to take too much time of the frame, so we reduced the resolution to 640x400. For a whole frame at 40MHz per half-byte this would take 15.36 ms, giving us only 1.3 ms per 60 Hz frame to do non-display work. At the smaller size it only takes 12.8 ms to emit a frame and 3.8 ms leftover time. Plus, 640x400 is widescreen, so it looks more modern!

Reviewer: ```suggestion
Lastly, even with the faster SPI pixel output, emitting the full 640x480 frame proved to take too much time, so we reduced the resolution to 640x400. For a whole frame at 40MHz per half-byte the full resolution would take 15.36 ms, giving us a budget of only 1.3 ms per 60 Hz frame for software overhead and other non-display work. At the smaller size it only takes 12.8 ms to emit a frame, tripling our leftover time to 3.8 ms per frame. Plus, 640x400 is widescreen, so it looks more modern!
```

With these pieces we could design a circuit that collects two 4 bit [nibbles](https://en.wikipedia.org/wiki/Nibble) into a byte (representing the color of a single pixel) and load it into a FIFO memory. At the same time, the VGA timing hardware would read pixel bytes exactly when needed.

### Logical Components

There are five main components to Vgaterm:

* VGA Timing
* CPU Pixel Output
* Displaying a Pixel Byte
* Digital to Analog Converters (DAC)
* IO control and video buffer updates

{% include figure.html image="/assets/images/vgaterm-logical-diagram.jpg" caption="Vgaterm logical diagram. Details were omitted to show the high level units" width=800 %}

Repositories containing all our code and schematics:
* esp32c3 code: [https://github.com/rustbox/esp32c3-vgaterm](https://github.com/rustbox/esp32c3-vgaterm)
* VGA timing generation code and Programmable Logic Device code: [https://github.com/rustbox/cuddly-robot](https://github.com/rustbox/cuddly-robot)
* PCB schematics and CAD: [https://github.com/rustbox/vgaterm-schematics](https://github.com/rustbox/vgaterm-schematics)

### VGA Timing

> VGA comes in several modes, and we target the 640x480, 60Hz frame rate mode - the base standard. VGA has 3 analog signals between 0 and 0.7 volts representing the amount of Red Green and Blue per pixel. There are also two timing signals `H_SYNC` and `V_SYNC` that tell the monitor what resolution mode to be in and when to sample the color signals to actually display a pixel. When the monitor is actively sampling colors and displaying pixels we say that this portion of the frame is `visible`. There is time during a complete frame when this isn't the case and this is the _blanking area_. Lastly, the _pixel clock_ speed is the rate at which pixels are shown in the visible portion, and in our case that's 25.175 MHz. In the diagram above, I've labeled a signal `CLK_VIS`, and this is the _pixel clock_ signal during the `visible` time. To see detailed VGA timing information, see [tiny vga](http://tinyvga.com/vga-timing/640x480@60Hz).

Reviewer: If you really want to introduce blanking area and define H_SYNC and V_SYNC (which, I don't think you need to), then I'd suggest taking it from the other end: why did we do the timing signals first?  I remember being really glad we did that, but the reader doesn't know why. What happened the handful of times something went wrong there?

Reviewer: FWIW, I still think this paragraph feels out-of-place: we were just doing high-level block diagrams, and now you're telling me about pixel clocks to 3 decimal places of precision? what gives? worse, it's the 2nd (actually 3rd) time the reader is encountering this detail, and there's not really a payoff for it: we never talk about other modes, or what would need to change to support them, so "pixel clock" is just as specific as "25.175MHz pixel clock" (for the purposes of this post, anyway).

Your next paragraph covers everything that needs to be said really well, I think, so I'd still suggest striking it out; failing that, maybe move it to the bottom of this section so it's not quite such an abrupt level-of-detail transition?

:----:|:-----:
![Timing Board front](/assets/images/vgaterm-timing-front.jpg) | ![Timing Board back](/assets/images/vgaterm-timing-back.jpg)
<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>

> VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.

Reviewer: doesn't matter b/c markdown (really: html) but

```suggestion
VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.
```

### Pixels Out

> Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. We decided that this is the most important software role as too much jitter in timing will cause artifacts and wobbling in the piture.

Reviewer: ```suggestion
Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. We decided that this is the most important software role as too much jitter in timing will cause artifacts and wobbling in the picture.
```

> To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
> 
> {% include figure.html image="/assets/images/fifo-chip-highlight.jpg" width=450 position="right" caption="IDT 7205 FIFO chip" %}
> 
> This worked well, but there are a few constraints we had to consider when designing a circuit around the FIFO:

Reviewer: ```suggestion
As mentioned, we used a FIFO chip to separate the SPI and VGA clock domains:

{% include figure.html image="/assets/images/fifo-chip-highlight.jpg" width=450 position="right" caption="IDT 7205 FIFO chip" %}

This worked well, but there are a few constraints we had to consider when designing a circuit around the FIFO:
```

* The FIFO read and write setup and hold times had to be obeyed by `CLK_VIS` and the SPI clock signals respectively
* The SPI only supported clock speeds in power-of-two divisions of 80 MHz (80, 40, 20, 10, etc)

> * If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels if the CPU didn't pause writing periodically.
> * If we emitted pixels too slowly, and VGA hardware consumed pixels faster than was being written, then we'd have to start loading pixels before VGA started consuming them.

Reviewer: ```suggestion
* If we emitted pixels too quickly, and VGA hardware did not read from the FIFO fast enough, we would blow the queue and drop pixels
* If we emitted pixels too slowly, and VGA hardware consumed pixels faster than was being written, then we would also be missing pixels
```

Pausing periodically was an effect, IIRC, of the DMA sizing and not a design goal: but loading pixels before the VGA started _was_ a design goal.

It's good to give the reader a sense of the difficulty in lining up the average read rates, but I don't think you need to go into the "20MHz, except when it wasn't" and "21.175MHz, but a whole bunch of the pixels are timing controlled and have to be black" details here.

> * The size of the FIFO effects how much wiggle room we get in the above two cases, especially as the FIFO is too expensive to have enough room for the whole frame.

Reviewer: Teknically yes, but the more common use of `effect` is as a noun; consider:

 ```suggestion
* The size of the FIFO affects how much wiggle room we get in the above two cases, especially as the FIFO is too expensive to have enough room for the whole frame.
```

Though this is a little redundant with the description I suggested earlier; since the FIFO was such a primary constraint I think it works (maybe with a callback?), but you could consider using foreshadowing above instead.

Reviewer: or even

```suggestion
* The FIFO is too expensive to have enough room for the whole frame, requiring more coordination between the readers & writers 
```

> * VGA hardware does not read from the FIFO at the pixel clock rate across the entire duration of the frame. Since the visible portion of the frame is only about 73% of the total time spent per frame (because of the blanking area), the average read rate per frame is actually less than the pixel clock rate

Reviewer: ```suggestion
* The average rate for the readers & writers across the frame was different than the instantaneous rate: both the VGA and SPI signals had pauses at different times (for different reasons)
```

If you want to go into (both) the reasons, that's cool, but that also feels like a really good thing to save for the future too.

> 

Reviewer: I think it's worth separating out the setup & result here: the bullets are introduced as "constraints" so I think it'd be best to keep them to strictly that, and then give the flavor of the solution here before the cliffhanger; something like:

```suggestion

<something something ended up with a design> where the (slower) SPI would plod ahead pre-filling a chunk of the buffer, and then the (faster, less consistent) would sprint to not-quite-catch-up <something something>

```

The details in how we eneded up on our specific design will be elaberated on in another post, but suffice to say we had to coordinate design details across software, hardware, specific part choices, and part pricing.

### Displaying Pixels

> To display pixels, the `CLK_VIS` signal generated by the VGA timing hardware reads bytes from the FIFO and into a [Programmable Logic Device (PLD)](https://en.wikipedia.org/wiki/Programmable_logic_device) which maps the 8 bits into 9 bits, 3 bits for each of the color channels - red, green, and blue. These 9 bits then get sent to the DAC, described in the following section.

Reviewer: ```suggestion
To display pixels, the `CLK_VIS` signal generated by the VGA timing hardware triggers reading bytes from the FIFO and into a [Programmable Logic Device (PLD)](https://en.wikipedia.org/wiki/Programmable_logic_device) which maps the 8 bits into 3 bits for each of the color channels - red, green, and blue. These 9 bits then get sent to the DAC, described in the following section.
```

two things:

1. it still feels really weird to me to describe clock signals as having agency; they're the clock on the wall that we all use to coordinate, not the thing that makes us do stuff. Also, did we literally end up wire the clock signal into /R on the FIFO? if so, that feels more like "clever optimization"/chance than agency, still
2. using a whole chip to map "8 bits into 9 bits" is a bit confusing as to why; saying "into 3 bits of each color" -> "oh that happens to be 9 bits" avoids the hiccup in the sentence

> An important consideration for this stage is how long it takes to shuttle a byte from the FIFO into the DAC and finally into the monitor. Since there is processing of the read bytes it's possible that the pixels become late compared to `H_SYNC` and `V_SYNC` and this will show up as a horizontal shift. I.E. the "first" pixel could actually show up on the 2nd or 3rd pixel position on the screen, instead of the actual 1st pixel.

Reviewer: ```suggestion
An important consideration for this stage is how long it takes to shuttle a byte from the FIFO into the DAC and finally into the monitor. The FIFO and color chip both have propagation delays that might make the read bytes show up "late," and contamination delays that might make them show up "early" compared to `H_SYNC` and `V_SYNC`. This will show up as a horizontal shift. I.E. the "first" pixel could actually show up on the 2nd or 3rd pixel position on the screen, instead of the actual 1st pixel.
```

{% include figure.html image="/assets/images/left-collum-timing-bug.jpg" caption="Notice the far left column of the screen in white next to the blue. It appears to wrap around from the far right side of the screen because the first pixel of the image is shifted over 2 or 3 positions" width=550 %}

> So we developed a method to handle this kind of pixel shifting to synchronize the pixel hardware with the VGA timing hardware.

Reviewer: say more? or less, I think: cutting this sentence would be fine too.

{% include figure.html image="/assets/images/vgaterm-dac-board.jpg" caption="R2R Resistor DAC" width=550 %}

Once the pixel byte has been taken out of the FIFO memory it's on its way to being displayed by the monitor. To be displayed this pixel byte must be converted into three analog voltages between 0 and 0.7 volts corresponding to the Red, Green, and Blue color channels. We accomplished this with an [R2R Resistor Ladder DAC](https://en.wikipedia.org/wiki/Resistor_ladder#R%E2%80%932R_resistor_ladder_network_(digital_to_analog_conversion)).

In an R2R resistor ladder DAC the resistor network is commonly connected to an op-amp to maintain the output voltage regardless of what load the DAC is connected to. However in our case the load is only ever the 75 Ω impedance of the analog color lines in the monitor, which we can account for in the resistor ladder design. One concern, however, is that our resistor ladder can consume a fair amount of current because we're going from 5 V input and dropping to under 1 V. The output pins representing the color coming from an IC may not be rated to supply the amount of current the DAC would need to maintain the voltage.

> In an earlier version of the DAC this is what we did (copying Ben Eater's video) but the colors will dull. We also tried an op-amp but that we saw some weird visual artifacts in the output. What seemed to work best was using a [line driver](https://www.ti.com/lit/gpn/SN74AHCT1G125) to supply the resister ladder with enough current to prevent the voltage dropping too low. Now the colors are properly vivid.

Reviewer: ```suggestion
In an earlier version of the DAC this is what we did anyway (copying Ben Eater's video), making the colors dull. We also tried an op-amp but we saw some weird visual artifacts in the output. What seemed to work best was using a [line driver](https://www.ti.com/lit/gpn/SN74AHCT1G125) to supply the resister ladder with enough current to prevent the voltage dropping too low. Now the colors are properly vivid.
```

{% include figure.html image="/assets/images/vgaterm-terminal-sequence.drawio.png" caption="Basic Keyboard input to terminal operation" %}

For keyboard input we went over several options including falling back on [PS/2 protocol over USB](https://en.wikipedia.org/wiki/PS/2_port#Conversion_between_PS/2_and_USB) and trying to write a basic software-driven USB host with the esp32c3 GPIO pins. But we settled on just using this [USB host to UART converter](https://www.tindie.com/products/matzelectronics/ch559-usb-host-to-uart-bridge-module/) by [MatzElectronics](https://www.tindie.com/stores/matzelectronics/).

Software, in a loop, waits for keyboard input or input by a connected host. When input is received on either, the CPU is interrupted. If receiving key presses, the character(s) are dequeued and passed along to the connected host. If receiving characters from the host, vgaterm updates the text buffer and cursor which updates the video memory. The updated video memory is then emitted in the next frame, showing the text to the user on screen.

> Seth and I decided that this should _never_ interrupt the CPU from emitting pixel bytes as we wanted to prevent any visual artifacts as much as possible. However this could mean it's possible to drop characters if the CPU is overwhelmed with input. We landed on this decision because the chance of dropping characters seemed small and the display shaking and wobbling felt viscerally "broken". We can also help mitigate the liklihood of dropping characters with some UART settings ([Software Flow Control](https://en.wikipedia.org/wiki/Software_flow_control)).

Reviewer: ```suggestion
Seth and I decided that this should _never_ interrupt the CPU from emitting pixel bytes as we wanted to prevent any visual artifacts as much as possible. However this could mean it's possible to drop characters if the CPU is overwhelmed with input. We landed on this decision because the chance of dropping characters seemed small and the display shaking and wobbling felt viscerally "broken". We can also help mitigate the likelihood of dropping characters with some UART settings ([Software Flow Control](https://en.wikipedia.org/wiki/Software_flow_control)).
```

# Closing Thoughts

This is only an overview of the operation and design of Vgaterm. My hope is that I can continue to make posts going into more detail about each of these major components. I learned so much working on this project and it gave me such joy to share it with a friend.

> Working on this project challenged me technologically at every step but I always felt like I could make progress. And when I did despair and feel overwhelmed by a problem I couldn't crack Seth would be there to help and together we'd make progress again. And likewise I would be there for Seth when he was stuck. For one of the first times I felt supported and heard in work. Our skills and outlooks complemented each other in a virtuous cycle. I could not have done this on my own, and together we made something and learned together which feels remarkable!

Reviewer: ```suggestion
Working on this project challenged me technologically at every step but I always felt like I could make progress. And when I did despair and feel overwhelmed by a problem I couldn't crack, Seth would be there to help and together we'd make progress again. And likewise I would be there for Seth when he was stuck. For one of the first times I felt supported and heard in work. Our skills and outlooks complemented each other in a virtuous cycle. I could not have done this on my own, and together we made something and learned together which feels remarkable!
```

Reviewer: also <3 <3

> I have more hope, after this experience, that I can find and cultivate more working relationships that are nurturing, where we learn from each other, and that allow us to tackle problems that we couldn't solve individually. Through community we can own and solve problems that colletively affect us.

Reviewer: ```suggestion
I have more hope, after this experience, that I can find and cultivate more working relationships that are nurturing, where we learn from each other, and that allow us to tackle problems that we couldn't solve individually. Through community we can own and solve problems that collectively affect us.
```

This leads to how I began to change how I thought about open source and open hardware as Vgaterm was developed. Seth and I always wanted this to be open, but now I see us and our project as existing within and for the open source community and also in the greater maker and DIY community. Communities like knitters, gardeners, crafters, artists, DIY, home chefs and others all use a network of engaged and creative folks to pass around ideas to share - to collectively own the ideas of their particlar craft for anyone to learn. I've started to see technology in a similar light.

> I've realized that electronics and hardware and programming are just skills that someone can learn like any other, and that we, the community, can collectively own and build things that are important and useful - that we don't have to capitulate to huge corporations that want to collect our data, exploit workers, or monopolize the objects and services in our daily lives. It feels like a way to fight against the [enshittification](https://en.wikipedia.org/wiki/Enshittification) that represents the current era in tech today. These are concepts I'm still thinking about and working through, and I have Vgaterm to thank for helping me engage with these ideas.

Reviewer: <3 <3 <3
